File: apps/client/src/xp.ts
const BASE_STAGE_THRESHOLDS = [0, 120, 280, 480, 720, 980, 1280, 1680];

export type StageProgress = {
  totalXP: number;
  stageIndex: number;
  currentStart: number;
  nextTarget: number;
  currentPercent: number;
  totalProgressPercent: number;
  nextLevelRemaining: number;
};

function getStageThresholds(totalStages: number): number[] {
  const safeTotalStages = Math.max(1, totalStages);
  const requiredLength = safeTotalStages + 1;
  const thresholds = [...BASE_STAGE_THRESHOLDS];

  while (thresholds.length < requiredLength) {
    const last = thresholds[thresholds.length - 1] ?? 0;
    thresholds.push(last + 400);
  }

  return thresholds.slice(0, requiredLength);
}

export function getStageRequirement(stageIndex: number, totalStages: number): number {
  const thresholds = getStageThresholds(totalStages);
  const safeIndex = Math.max(0, Math.min(stageIndex, thresholds.length - 2));
  return thresholds[safeIndex] ?? 0;
}

export function calculateStageProgress(totalXP: number, totalStages: number): StageProgress {
  const thresholds = getStageThresholds(totalStages);
  const safeTotalXP = Math.max(0, totalXP);
  const lastStageIndex = thresholds.length - 2;
  let stageIndex = lastStageIndex;

  for (let index = 0; index <= lastStageIndex; index += 1) {
    if (safeTotalXP < thresholds[index + 1]) {
      stageIndex = index;
      break;
    }
  }

  const currentStart = thresholds[stageIndex] ?? 0;
  const nextTarget = thresholds[stageIndex + 1] ?? currentStart;
  const denominator = Math.max(1, nextTarget - currentStart);
  const finalMilestone = thresholds[thresholds.length - 1] ?? 1;

  return {
    totalXP: safeTotalXP,
    stageIndex,
    currentStart,
    nextTarget,
    currentPercent: Math.min(100, Math.max(0, Math.round(((safeTotalXP - currentStart) / denominator) * 100))),
    totalProgressPercent: Math.min(100, Math.max(0, Math.round((safeTotalXP / Math.max(1, finalMilestone)) * 100))),
    nextLevelRemaining: Math.max(0, nextTarget - safeTotalXP)
  };
}

export const MAX_LEARNING_LEVEL = 999;

export type LearningProgress = {
  level: number;
  xpIntoLevel: number;
  xpToNextLevel: number;
  levelProgress: number;
  rankCode: string;
  rankLabel: string;
};

function xpForNextLevel(level: number): number {
  if (level < 100) return 6 + Math.floor((level - 1) / 25);
  if (level < 300) return 10 + Math.floor((level - 100) / 20);
  if (level < 600) return 20 + Math.floor((level - 300) / 18);
  return 38 + Math.floor((level - 600) / 16);
}

export function getLearningRank(level: number): Pick<LearningProgress, "rankCode" | "rankLabel"> {
  if (level >= 999) return { rankCode: "LEGEND", rankLabel: "Lingua 传说" };
  if (level >= 900) return { rankCode: "SOVEREIGN", rankLabel: "至尊领航者" };
  if (level >= 800) return { rankCode: "CELESTIAL", rankLabel: "星耀宗师" };
  if (level >= 650) return { rankCode: "MASTER", rankLabel: "语言大师" };
  if (level >= 500) return { rankCode: "EXPERT", rankLabel: "表达专家" };
  if (level >= 350) return { rankCode: "SCHOLAR", rankLabel: "语境学者" };
  if (level >= 200) return { rankCode: "ADEPT", rankLabel: "进阶研习者" };
  if (level >= 100) return { rankCode: "VOYAGER", rankLabel: "表达旅人" };
  if (level >= 50) return { rankCode: "EXPLORER", rankLabel: "语言行者" };
  return { rankCode: "NOVICE", rankLabel: "初学探索者" };
}

export function calculateLearningProgress(totalXP: number): LearningProgress {
  let remaining = Math.max(0, totalXP);
  for (let level = 1; level < MAX_LEARNING_LEVEL; level += 1) {
    const required = xpForNextLevel(level);
    if (remaining < required) {
      return { level, xpIntoLevel: remaining, xpToNextLevel: required, levelProgress: Math.floor((remaining * 100) / required), ...getLearningRank(level) };
    }
    remaining -= required;
  }
  return { level: MAX_LEARNING_LEVEL, xpIntoLevel: 0, xpToNextLevel: 0, levelProgress: 100, ...getLearningRank(MAX_LEARNING_LEVEL) };
}

File: apps/server/internal/service/xp.go
package service

import (
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/linguaquest/server/internal/domain"
)

const dailyXPCap = 300

type XPStore interface {
	AwardXP(event domain.XPEvent, dailyCap int, dayStart time.Time) (domain.XPAward, error)
	ListXPEvents(userID string, limit int) ([]domain.XPEvent, error)
}

func (s *Service) awardLearningXP(userID string, activity string, sourceID string, quality int) (int, error) {
	if userID == "" || sourceID == "" {
		return 0, errors.New("missing XP event identity")
	}
	event := domain.XPEvent{ID: uuid.NewString(), UserID: userID, Activity: activity, SourceID: sourceID, XPEarned: learningXPAmount(activity, quality), CreatedAt: time.Now().In(time.FixedZone("CST", 8*60*60))}
	if store, ok := s.store.(XPStore); ok {
		dayStart := time.Date(event.CreatedAt.Year(), event.CreatedAt.Month(), event.CreatedAt.Day(), 0, 0, 0, 0, event.CreatedAt.Location())
		award, err := store.AwardXP(event, dailyXPCap, dayStart)
		if err != nil {
			return 0, err
		}
		return award.GrantedXP, nil
	}
	if err := s.store.AddUserXP(userID, event.XPEarned); err != nil {
		return 0, err
	}
	return event.XPEarned, nil
}

func (s *Service) XPEvents(userID string, limit int) ([]domain.XPEvent, error) {
	store, ok := s.store.(XPStore)
	if !ok {
		return []domain.XPEvent{}, nil
	}
	if limit <= 0 || limit > 30 {
		limit = 12
	}
	return store.ListXPEvents(userID, limit)
}

func learningXPAmount(activity string, quality int) int {
	if quality < 0 {
		quality = 0
	}
	if quality > 100 {
		quality = 100
	}
	switch activity {
	case "THEATER_PRACTICE":
		return 10 + quality*22/100
	case "READING_PRACTICE":
		return 10 + quality*24/100
	case "ROLEPLAY_COMPLETE":
		return 16 + quality*30/100
	case "WRITING_COMPLETE":
		return 20 + quality*40/100
	case "SPEAKING_TURN":
		return 8 + quality*18/100
	default:
		return 8
	}
}

func learningProgress(totalXP int) domain.LearningProgress {
	if totalXP < 0 {
		totalXP = 0
	}
	remaining := totalXP
	for level := 1; level < domain.MaxLevel; level++ {
		required := xpForNextLevel(level)
		if remaining < required {
			return domain.LearningProgress{Level: level, XPIntoLevel: remaining, XPToNextLevel: required, LevelProgress: remaining * 100 / required, RankCode: rankForLevel(level), RankLabel: rankLabelForLevel(level)}
		}
		remaining -= required
	}
	return domain.LearningProgress{Level: domain.MaxLevel, XPIntoLevel: 0, XPToNextLevel: 0, LevelProgress: 100, RankCode: "LEGEND", RankLabel: "Lingua 传说"}
}

func xpForNextLevel(level int) int {
	switch {
	case level < 100:
		return 6 + (level-1)/25
	case level < 300:
		return 10 + (level-100)/20
	case level < 600:
		return 20 + (level-300)/18
	default:
		return 38 + (level-600)/16
	}
}

func rankForLevel(level int) string {
	switch {
	case level >= 999:
		return "LEGEND"
	case level >= 900:
		return "SOVEREIGN"
	case level >= 800:
		return "CELESTIAL"
	case level >= 650:
		return "MASTER"
	case level >= 500:
		return "EXPERT"
	case level >= 350:
		return "SCHOLAR"
	case level >= 200:
		return "ADEPT"
	case level >= 100:
		return "VOYAGER"
	case level >= 50:
		return "EXPLORER"
	default:
		return "NOVICE"
	}
}

func rankLabelForLevel(level int) string {
	labels := map[string]string{
		"NOVICE": "初学探索者", "EXPLORER": "语言行者", "VOYAGER": "表达旅人", "ADEPT": "进阶研习者", "SCHOLAR": "语境学者", "EXPERT": "表达专家", "MASTER": "语言大师", "CELESTIAL": "星耀宗师", "SOVEREIGN": "至尊领航者", "LEGEND": "Lingua 传说",
	}
	return labels[rankForLevel(level)]
}

func decorateLearningProgress(user domain.User) domain.User {
	progress := learningProgress(user.TotalXP)
	user.Level = progress.Level
	user.XPIntoLevel = progress.XPIntoLevel
	user.XPToNextLevel = progress.XPToNextLevel
	user.LevelProgress = progress.LevelProgress
	user.RankCode = progress.RankCode
	user.RankLabel = progress.RankLabel
	return user
}
